docs/development/fleetlock/protocol.md

```diff
@@ -0,0 +1,95 @@
+
+When a client starts it proceeds into either "initialization" or "finalization" state, based on system condition (i.e. whether a finalization is already in progress).
+In "**initialization**" state, the client tries to release any reboot slot it may have previously held.
+A successful unlock operation means that the client can proceed into its "**steady**" state and look for further updates.
+When an update is found and locally staged, the client proceed into its "**pre-reboot**" state and tries to lock a reboot slot.
+A successful lock operation means that the client can proceed into its "**finalization**" state and finalize a pending update, then reboot.
+
+## Requests
+
+### Endpoints
+
+All endpoints defined below are relative to a common deployment-specific base URL:
+
+ * `/v1/pre-reboot`: reserve/lock a reboot slot
+ * `/v1/steady-state`: release/unlock a reboot slot
+
+### Body
+
+All POST requests contain well-formed JSON body according to the following schema:
+
+ * `client_params` (object, mandatory)
+   * `id` (string, mandatory, non-empty): client identifier (e.g. node name or UUID)
+   * `group` (string, mandatory, non-empty): reboot-group of the client
+
+Client ID is a case-sensitive textual label that uniquely identifies a lock holder. It is generated and persisted by each client.
+Client group is a textual label conforming to the regexp `^[a-zA-Z0-9.-]+$`. The recommended default group name is "`default`".
+
+### Headers
+
+Locking and unlocking requests must contain a `fleet-lock-protocol` header with a fixed value of `true` to ensure that the actual request was directly intended and not a part of unintentional redirection.
+
+### Response
+
+If the operation is succesful, a 200 status code is returned. Every other code is considered as a failed operation.
+
+### Example
+
+A client with UUID `c988d2509fdf4cdcbed39037c56406fb` and group `workers` can try to acquire a reboot slot from `https://example.com/base` in a way which is conceptually similar to the following:
+
+Request body:
+
+```json
+
+{
+  "client_params": {
+    "group": "default",
+    "id": "c988d2509fdf5cdcbed39037c56406fb"
+  }
+}
+
+```
+
+POST request:
+
+```shell
+
+curl -H "fleet-lock-protocol: true" -d @body.json http://example.com/base/v1/pre-reboot
+
+```
+
+### Errors
+
+Errors on the service endpoints SHOULD be returned to the client as JSON objects, with a 4xx or 5xx HTTP status code.
+Error values carry a type-identifier and a textual description, according to the following schema:
+
+|  Key   | Optional | Description                                                  |
+|--------|----------|--------------------------------------------------------------|
+| kind   | required | error type identifier, as a non-empty JSON string            |
+| value  | required | human-friendly error description, as a non-empty JSON string |
+
+This allows clients to show more specific error details to cluster administrators, instead of generic HTTP errors.
```

Comment:
```suggestion
This allows clients to show more specific error details to cluster administrators, instead of generic HTTP errors.

Example:

```json

{
    "kind": "generic_http_500", 
    "value":  "(unknown/generic server error)"
}
```

Reply:
I think showing an example is helpful but ... I don't think I picked the best example :sweat_smile:.

Reply:
Good point. I'll add an example for "all slots already locked" negative response, which is a real case with more specific details.